File: src/service-maker/sources/core/src/source_config.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "source_config.hpp"

using namespace deepstream;

#define KEY_SOURCE_LIST    "source-list"
#define KEY_SOURCE_CONFIG  "source-config"
#define KEY_SOURCE_CONFIG_BIN "source-bin"
#define KEY_SOURCE_CONFIG_PROPERTIES "properties"

#define KEY_CAMERA_LIST "camera-list"

SourceConfig::SourceConfig(const std::string& config_file) {
  YAML::Node root = YAML::LoadFile(config_file);
  for (YAML::const_iterator it1 = root.begin(); it1 != root.end(); ++it1) {
    std::string key = it1->first.as<std::string>();
    if (key == KEY_SOURCE_LIST) {
      YAML::Node list_node = it1->second;
      for (YAML::const_iterator it2 = list_node.begin(); it2 != list_node.end(); ++it2) {
        SensorInfo sensor_info = {"N/A", "N/A", "N/A"};
        const YAML::Node& sensor_node = *it2;
        for (YAML::const_iterator it3 = sensor_node.begin(); it3 != sensor_node.end(); ++it3) {
          std::string name = it3->first.as<std::string>();
          std::string value = it3->second.as<std::string>();
          if (name == "uri") {
            sensor_info.uri = value;
          } else if (name == "sensor-id") {
            sensor_info.sensor_id = value;
          } else if (name == "sensor-name") {
            sensor_info.sensor_name = value;
          }
          else
          {
            std::string what = "Invalid key in source config: ";
            throw std::runtime_error(what + name);
          }
        }
        sensor_info_.push_back(sensor_info);
      }
    }
    else if (key == KEY_CAMERA_LIST)
    {
      use_camerabin_ = true;
      YAML::Node list_node = it1->second;
      for (YAML::const_iterator it2 = list_node.begin(); it2 != list_node.end(); ++it2)
      {
        CameraInfo camera_info;
        const YAML::Node& sensor_node = *it2;
        for (YAML::const_iterator it3 = sensor_node.begin(); it3 != sensor_node.end(); ++it3)
        {
          std::string name = it3->first.as<std::string>();
          std::string value = it3->second.as<std::string>();
          if (name == "camera-type")
          {
            camera_info.camera_type = value;
          }
          else if (name == "camera-v4l2-dev-node")
          {
            camera_info.camera_v4l2_dev_node = value;
          }
          else if (name == "camera-csi-sensor-id")
          {
            camera_info.camera_csi_sensor_id = std::stoi(value);
          }
          else if (name == "camera-width")
          {
            camera_info.camera_width = value;
          }
          else if (name == "camera-height")
          {
            camera_info.camera_height = value;
          }
          else if (name == "camera-fps-n")
          {
            camera_info.camera_fps_n = value;
          }
          else if (name == "camera-fps-d")
          {
            camera_info.camera_fps_d = value;
          }
          else if (name == "gpu-id")
          {
            camera_info.gpu_id = std::stoi(value);
          }
          else if (name == "nvbuf-mem-type")
          {
            camera_info.nvbuf_mem_type = std::stoi(value);
          }
          else if (name == "nvvideoconvert-copy-hw")
          {
            camera_info.nvvideoconvert_copy_hw = std::stoi(value);
          }
          else
          {
            std::string what = "Invalid key in source config: ";
            throw std::runtime_error(what + name);
          }
        }
        camera_info_.push_back(camera_info);
      }
    }
    else if (key == KEY_SOURCE_CONFIG)
    {
      YAML::Node config_node = it1->second;
      for (YAML::const_iterator it2 = config_node.begin(); it2 != config_node.end(); ++it2)
      {
        std::string name = it2->first.as<std::string>();
        if (name == KEY_SOURCE_CONFIG_BIN)
        {
          use_nvmultiurisrcbin_ = it2->second.as<std::string>() == "nvmultiurisrcbin";
          use_nvurisrcbin_ = it2->second.as<std::string>() == "nvurisrcbin";
        }
        else if (name == KEY_SOURCE_CONFIG_PROPERTIES)
        {
          properties_ = it2->second;
        }
        else
        {
          std::string what = "Invalid key in source config: ";
          throw std::runtime_error(what + name);
        }
      }
    }
  }
}

uint32_t SourceConfig::nSources() const {
  return sensor_info_.size();
}

uint32_t SourceConfig::nCameraSources() const
{
  return camera_info_.size();
}

std::string SourceConfig::listSensorIds() const {
  std::string str = "";
  size_t n = sensor_info_.size();
  for (size_t i = 0; i < n; i++) {
    str += sensor_info_[i].sensor_id;
    if (i < n - 1) {
        str += ",";
    }
  }
  return str;
}

std::string SourceConfig::listSensorNames() const {
  std::string str = "";
  size_t n = sensor_info_.size();
  for (size_t i = 0; i < n; i++) {
    str += sensor_info_[i].sensor_name;
    if (i < n - 1) {
        str += ",";
    }
  }
  return str;
}

std::string SourceConfig::listUris() const {
  std::string str = "";
  size_t n = sensor_info_.size();
  for (size_t i = 0; i < n; i++) {
    str += sensor_info_[i].uri;
    if (i < n - 1) {
        str += ",";
    }
  }
  return str;
}

SensorInfo SourceConfig::getSensorInfo(uint32_t index) const {
  return sensor_info_[index];
}

CameraInfo SourceConfig::getCameraInfo(uint32_t index) const
{
  return camera_info_[index];
}

const YAML::Node& SourceConfig::getProperties() const {
  return properties_;
}

bool SourceConfig::useMultiUriSrcBin() const {
  return use_nvmultiurisrcbin_;
}

bool SourceConfig::useUriSrcBin() const
{
  return use_nvurisrcbin_;
}

bool SourceConfig::useCameraBin() const
{
  return use_camerabin_;
}

File: src/service-maker/sources/python/src/pyservicemaker/source_config.py
# SPDX-FileCopyrightText: Copyright (c) 2024-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import yaml
from collections import namedtuple
from dataclasses import dataclass
from typing import Optional

SensorInfo = namedtuple("SensorInfo", ["sensor_id", "sensor_name", "uri"])
CameraInfo = namedtuple("CameraInfo", ["camera_type", "camera_video_format", "camera_width", "camera_height", "camera_fps_n", "camera_fps_d",
                                       "camera_csi_sensor_id", "camera_v4l2_dev_node", "gpu_id", "nvbuf_mem_type",
                                       "nvvideoconvert_copy_hw"])

class SourceConfig:
    """Class for parsing source configuration"""
    def __init__(self):
        self._sensor_list = []
        self._camera_list = []
        self._source_type = None
        self._source_properties = dict()

    @property
    def sensor_list(self):
        return self._sensor_list

    @property
    def camera_list(self):
        return self._camera_list

    @property
    def source_type(self):
        return self._source_type

    @property
    def source_properties(self):
        return self._source_properties

    def load(self, config_file: str):
        """Load source configurations from a YAML file"""
        with open(config_file, 'r') as file:
            doc = yaml.safe_load(file)
            if doc:
                source_list = doc.get("source-list")
                camera_list = doc.get("camera-list")
                source_config = doc.get("source-config")
                if source_list is not None:
                    for source in source_list:
                        uri = source.get("uri")
                        sensor_id = source.get("sensor-id")
                        sensor_name = source.get("sensor-name")
                        self._sensor_list.append(SensorInfo(sensor_id, sensor_name, uri))
                if camera_list is not None:
                    for camera in camera_list:
                        camera_type = camera.get("camera-type")
                        camera_video_format = camera.get("camera-video-format")
                        camera_width = camera.get("camera-width")
                        camera_height = camera.get("camera-height")
                        camera_fps_n = camera.get("camera-fps-n")
                        camera_fps_d = camera.get("camera-fps-d")
                        camera_csi_sensor_id = camera.get("camera-csi-sensor-id")
                        camera_v4l2_dev_node = camera.get("camera-v4l2-dev-node")
                        gpu_id = camera.get("gpu-id")
                        nvbuf_mem_type = camera.get("nvbuf-mem-type")
                        nvvideoconvert_copy_hw = camera.get("nvvideoconvert-copy-hw")
                        self._camera_list.append(CameraInfo(camera_type, camera_video_format, camera_width, camera_height, camera_fps_n, camera_fps_d,
                                                            camera_csi_sensor_id, camera_v4l2_dev_node, gpu_id, nvbuf_mem_type,
                                                            nvvideoconvert_copy_hw))
                if source_config:
                    self._source_type = source_config.get("source-bin")
                    self._source_properties = source_config.get("properties")
                if camera_list is not None and len(camera_list):
                    self._source_type = "camerabin"


@dataclass
class RecordConfig:
    """Unified recording configuration for DeepStream pipeline.

    Set ``recording_type`` to either "local" or "cloud".

    - When "cloud": provide broker/protocol settings (Kafka, etc.).
    - When "local": only local recording settings are needed.

    Cloud parameters (only used when recording_type == "cloud"):
        - proto_lib: Path to protocol library (e.g., Kafka proto lib)
        - conn_str: Connection string (e.g., "localhost;9092")
        - msgconv_config_file: Message converter config file path
        - proto_config_file: Protocol adaptor config file path
        - topic_list: Comma-separated list of topics

    Common recording parameters (applied to smart-rec properties):
        - rec_cache: Cache size in seconds (default 20)
        - rec_container: Container format (0: MP4, 1: MKV)
        - rec_dir_path: Output directory path
        - rec_mode: 0 both, 1 video-only, 2 audio-only
    """
    recording_type: str = "local"  # "local" or "cloud"

    # Cloud-only fields
    proto_lib: Optional[str] = None
    conn_str: Optional[str] = None
    msgconv_config_file: Optional[str] = None
    proto_config_file: Optional[str] = None
    topic_list: Optional[str] = None

    # Common recording fields
    rec_cache: int = 20
    rec_container: int = 0
    rec_dir_path: str = "."
    rec_mode: int = 0

    def __post_init__(self):
        # Enforce mandatory cloud properties when recording_type is 'cloud'
        if isinstance(self.recording_type, str) and self.recording_type.lower() == "cloud":
            required = [
                "proto_lib",
                "conn_str",
                "msgconv_config_file",
                "proto_config_file",
                "topic_list",
            ]
            missing = [name for name in required if not getattr(self, name)]
            if missing:
                raise ValueError(f"Missing required cloud recording fields: {', '.join(missing)}")